state-sync/state-sync-v2/state-sync-driver/src/bootstrapper.rs

```diff
@@ -462,6 +473,54 @@ impl<
         }
     }
 
+    /// Fetches all missing state snapshot data in order to bootstrap the node
+    async fn fetch_missing_state_snapshot_data(
+        &mut self,
+        highest_synced_version: Version,
+        highest_known_ledger_info: LedgerInfoWithSignatures,
+    ) -> Result<(), Error> {
+        if highest_synced_version == GENESIS_TRANSACTION_VERSION {
+            // We're syncing a new node. Check the progress and fetch the missing data.
+            if let Some(target) = self.metadata_storage.previous_snapshot_sync_target()? {
+                if self.metadata_storage.is_snapshot_sync_complete(&target)? {
+                    panic!(
+                        "The snapshot sync for the target was marked as complete but \
+                    the highest synced version is genesis! Something has gone wrong! \
+                    Target snapshot sync: {:?}",
+                        target
+                    );
+                } else {
+                    self.fetch_missing_state_values(target, true).await
+                }
+            } else {
+                // No snapshot sync has started. Start a new sync for the highest known ledger info.
+                self.fetch_missing_state_values(highest_known_ledger_info, false)
+                    .await
+            }
+        } else {
+            // This node has already synced some state. Ensure the node is not too far behind.
+            let highest_known_ledger_version = highest_known_ledger_info.ledger_info().version();
+            let num_versions_behind = highest_known_ledger_version
+                .checked_sub(highest_synced_version)
+                .ok_or_else(|| {
+                    Error::IntegerOverflow("The number of versions behind has overflown!".into())
+                })?;
+            if num_versions_behind
+                < self
+                    .driver_configuration
+                    .config
+                    .num_versions_to_skip_snapshot_sync
+            {
+                self.bootstrapping_complete() // Skip the state snapshot sync
+            } else {
+                panic!("Snapshot syncing is currently unsupported for nodes with existing state! \
```

Comment:
Yeah, we don't want gaps in data (because it'll screw a lot of things up). 

> If the concern is leaving a hole in the DB, we can just delete the DB here and continue?

Regarding how to handle this, I think panicking is the most eloquent: the user has instructions on what to do and how to respond. We can also add docs, etc. Choosing to automatically delete all the data does not sound like a good experience, e.g., imagine if a node operator who has a lot of data synced in the database accidentally changes the config and starts up. Should state sync then go and wipe the entire DB, just to start syncing? That sounds very scary to me 😱 I'd much rather the user make this decision 😄

Reply:
I am not sure if its the right thing to do here. Given the current configuration of ensure we are at most 1M version behind and 1K TPS network, if a node is down for 30 mins, its never going to be back without a manual intervention. If we don't want to automatically delete the DB, what are the thoughts on making this window longer? So that nodes that are down for several hours can come back and catch up with the network?

Reply:
Synced offline and bumped the default to 10M, which provides around 3 hours 😄 We can always revisit it.